### PR TITLE
plug in InitUserMeshBlockData

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

As discussed in the call today and on the Matrix channel, there's a need for initializing variables and setting grid quantities once per re-meshing, but not more frequently than that. To enable this, I re-enable `InitUserMeshBlockData`, which is set in the application input object. This is now called at the beginning of `Mesh::Initialize` before anything else. `Mesh::Initialize` is called at the beginning of a simulation and upon each remeshing. This means data set in this function is expected to remain constant except when remeshing, and is available to problem generators, boundary conditions, and `FillDerived`.

I also hook it in to the `calculate_pi` example to show it works.

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [x] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

**Comment by Yurlungur** [comment]
@par-hermes format

**Comment by Yurlungur** [comment]
I have checked that this fits our needs downstream.

**Review by brryan — COMMENTED** [review]
Looks good to me. Only comment is would it be worth mentioning this usage of `InitUserMeshBlockData` in e.g. `~/github/parthenon/docs/amr.md`? From the existing documentation it's not immediately apparent (to me at least) that `InitUserMeshBlockData` is called after every remeshing step.

**Comment by Yurlungur** [comment]
Oh good call, @brryan . Yes documentation should be added.

File: docs/amr.md [hunk 1]
```diff
@@ -34,3 +34,21 @@ In each refinement block, you are required to provide a ``method`` which is a st
 
 ## Package-specific Criteria
 As a package developer, you can define a tagging function that takes a ``Container`` as an argument and returns an integer in {-1,0,1} to indicate the block should be derefined, left alone, or refined, respectively.  This function should be registered in a ``StateDescriptor`` object by assigning the ``CheckRefinement`` function pointer to point at the packages function.  An example is demonstrated [here](../example/calculate_pi/pi.cpp).
+
+## Ensuring your data is consistent after re-meshing
+
+WHen re-meshing happens, a few operations have, which can be plugged in to in various ways. The operations performed (in order) are:
```
Comment (brryan) [inline docs/amr.md]: `have` -> `happen`. I'm being picky, sorry

Comment (Yurlungur) [reply docs/amr.md]: Fixed. :+1:

**Review by brryan — APPROVED** [review]

File: src/application_input.hpp [hunk 2]
```diff
@@ -52,7 +52,7 @@ struct ApplicationInput {
   // MeshBlock functions
   std::function<std::unique_ptr<MeshBlockApplicationData>(MeshBlock *, ParameterInput *)>
       InitApplicationMeshBlockData = nullptr;
-  std::function<void(ParameterInput *)> InitUserMeshBlockData = nullptr;
+  std::function<void(MeshBlock *, ParameterInput *)> InitUserMeshBlockData = nullptr;
```
Comment (jlippuner) [inline src/application_input.hpp]: I think passing a pointer to the `MeshBlock` is the right thing to do here. But because we also have a `MeshBlockData` class, the name of the function might be a bit confusing (since it's user data for the `MeshBlock` and not the `MeshBlockData`). What about renaming the function to `InitMeshBlockUserData`?

**Review by jlippuner — APPROVED** [review]
Looks great! I just have a small suggestion to rename the function to avoid potential confusion with `MeshBlockData`, but I'll defer to your judgement.

File: src/application_input.hpp (hunk 2 above)
Comment (Yurlungur) [reply src/application_input.hpp]: I like this idea. I'll make the change.
